File: Algorithms/IntersectionofTwoLinkedLists/IntersectionofTwoLinkedLists.py
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        get_length = self.get_length
        move_eq = self.move_to_equal
        
        ptr_A, ptr_B = headA, headB
        len_A = get_length(headA)
        len_B = get_length(headB)
        
        if len_A > len_B:
            ptr_A = move_eq(ptr_A, len_A-len_B)
        else:
            ptr_B = move_eq(ptr_B, len_B-len_A)
        
        while ptr_A is not None:
            if ptr_A == ptr_B:
                return ptr_A
            ptr_A = ptr_A.next
            ptr_B = ptr_B.next
        return None
        
    def move_to_equal(self, head, steps):
        ptr = head
        for _ in range(steps):
            ptr = ptr.next
        return ptr
    
    def get_length(self, head):
        length = 0
        ptr = head
        while ptr is not None:
            length += 1
            ptr = ptr.next
        return length

File: Algorithms/IntersectionofTwoLinkedLists/IntersectionofTwoLinkedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB){
        int len_A = size(headA);
        int len_B = size(headB);
        int diff = len_A - len_B;
        
        if(diff > 0){
            while(diff > 0){
                headA = headA->next;
                diff--;
            }
        }
        else{
            while (diff < 0){
                headB = headB->next;
                diff++;
            }
        }
        
        while(headA){
            if(headA == headB) return headA;
            headA = headA->next;
            headB = headB->next;
        }
        return NULL;
    }
    
    int size(ListNode* node){
        if(node == NULL) return 0;
        return size(node->next) + 1;
    }
};